### ISSUE TITLE
When used in multiple thread, all the progress bars are in one line

### ISSUE THREAD

**Issue by chaoyangnz** [ctongfei/progressbar#11]
I use a progressbar for each thread, but all the progress bars are shown in one single line.

**Comment by ctongfei** [comment]
Could you please clarify? Are you using a progressbar for each thread, and those threads are running in parallel? 
If that is the case, then the current version of progressbar does not support that. Writing multiple progressbars to the console simultaneously require tweaking the cursor location in the terminal window, which is an issue that is not yet completely resolved.

**Comment by richdyang** [comment]
Yes, correctly. I use multiple threads and one thread one progress bar.

**Comment by ctongfei** [comment]
This is not currently doable right now. Will probably be fixed in a later version.

**Comment by Veske** [comment]
Yeah, this would be a quite useful feature for the project.

**Comment by ctongfei** [comment]
This involves the manipulation of cursor position in the terminal window which is a bit hard. Will probably be  fixed in a later version (1.0?)

**Comment by berlinguyinca** [comment]
any update on this feature?

**Comment by ctongfei** [comment]
It is on the horizon but recently I've got no time to implement this.

**Comment by aleck7** [comment]
As an option you can output multiple threads in one line, i.e.:
Map 10% [==>           ]  644/5855 (0:02:50 / 0:22:55) Reduce 2% [>           ] 123/5855 (0:02:50 / 0:22:55)
or even simpler:
Map|Reduce|Save [10%==>          |2%>            |0%             ]  817/17645 (0:02:50 / 0:22:55)

Can elaborate on details of elapsed/estimated time.

**Comment by dodalovic** [comment]
@ctongfei Is an instance of `ProgressBar` thread safe?

**Comment by ctongfei** [comment]
@dodalovic It is thread safe after version `0.5.0`.

**Comment by o2themar** [comment]
So now I see what you are talking about. The picture in my previous comment was when I had the progress bar delegate to a logger and I had the logger printing stuff to the console. After removing the consumer I now see the progress bar overwriting each other on one line. However they create a new line when they update. I also see some repeating going on the right. Not sure what that is all about. Look at the image below for more detail.
![image](https://user-images.githubusercontent.com/5449024/78358327-01d21e80-7581-11ea-8635-e6e8daf35e4b.png)

**Comment by ctongfei** [comment]
@o2themar The planned solution is to have a multi-line progressbar, where each line prints the progress of one thread. This is hard to implement --- sorry for not supporting this. I'll consider implementing this when I'm not busy, or you could start a PR!

**Comment by vehovsky** [comment]
The problem with current implementation is simple actually

> ConsoleProgressBarConsumer
```Java
out.print('\r'); // before update
out.print(str);
```
What _print('\r')_ does is that it moves the _cursor_ to the beginning of the line. And then the line is rewritten.


Based [ANSI/VT100 Terminal Control Escape Sequences](http://www.termsys.demon.co.uk/vtansi.htm) to implement multiline progressbar you would have to be moving the _cursor_ up and down to rewrite the corresponding line in the output.

Something like (over simplified obviously):

```Kotlin
var cursorEnd = 1

class TerminalPrinter(initialText: String) {
    init {
        if (System.console() == null) throw RuntimeException("non-interactive console")
        System.out.print('\r')
        System.out.println(initialText)
    }

    private val position = cursorEnd++

    suspend fun rewrite(str: String) {
        delay(500)
        moveCursorUp(cursorEnd - position)
        replaceLine(str)
        moveCursorDown(cursorEnd - position)
    }

    private fun moveCursorUp(count: Int) {
        System.out.print("\u001b[${count}A")
    }

    private fun moveCursorDown(count: Int) {
        System.out.print("\u001b[${count}B")
    }

    private fun replaceLine(str: String) {
        System.out.print('\r')
        System.out.print(str)
        System.out.print('\r')
    }
}

fun main() = runBlocking {
    val a = TerminalPrinter("a")
    val b = TerminalPrinter("b")
    val c = TerminalPrinter("c")
    val d = TerminalPrinter("d")

    a.rewrite("aa")
    c.rewrite("cc")
    b.rewrite("bb")
    val e = TerminalPrinter("e")
    d.rewrite("dd")
    e.rewrite("ee")

    a.rewrite("aaa")
    e.rewrite("eee")
    b.rewrite("bbb")
    d.rewrite("ddd")
    c.rewrite("ccc")
}
```

**Comment by vslee** [comment]
@vehovsky, can you send a pull request?

**Comment by vehovsky** [comment]
@vslee Not really, the above example works well on osx (linux as well probably) but not on windows. Moving the cursor just up/down (without asking terminal for cursor position) would probably work on windows as well, but that is just way too naive implementation and would break a minute some other thread wrote anything to console.

**Comment by ctongfei** [comment]
@vehovsky I understand these ANSI control code (and will probably have to be tested under many different envs, and use `jline` to actually check the capabilities of the terminal). Yes this is a planned feature for the next release (But I don't know when that will be).

**Comment by vehovsky** [comment]
I've prepared some initial implementation. Will create PR (after work) to start discussion about implementation details.

**Comment by vehovsky** [comment]
@o2themar 
>  was wondering though if it was possible to get the progress bar to print on 1 line rather than it keeps appending to the logger.
> ![image](https://user-images.githubusercontent.com/5449024/78298693-c7c13800-7500-11ea-9822-0ab9992055fc.png)

Yes it is possible. Not supported out of the box thought.

So to summarise the problem here. Logger will use defined pattern, lets use simple logback pattern as example:

```
[%d{YYYY-MM-dd HH:mm:ss.SSS}]%msg%n
```
Basically here our log message (_%msg_) will be prefixed with timestamp and suffixed with new line.

The problem is that potentially the control codes to move cursor are in the message (in the middle of your pattern). Also I said potentially, because DelegatingProgressBarConsumer does not send any.

So you would first have to write your own consumer. Something like:

```java
public class CustomDelegatingProgressBarConsumer implements ProgressBarConsumer {

    private final Character MOVE_CURSOR_TO_LINE_START = '\r';
    private final int maxProgressLength;
    private final Consumer<String> consumer;

    public CustomDelegatingProgressBarConsumer(Consumer<String> consumer, int maxProgressLength) {
        this.maxProgressLength = maxProgressLength;
        this.consumer = consumer;
    }

    @Override
    public int getMaxProgressLength() {
        return maxProgressLength;
    }

    @Override
    public void accept(String str) {
        this.consumer.accept(MOVE_CURSOR_TO_LINE_START + str);
    }

    @Override
    public void close() {
        this.consumer.accept(System.lineSeparator());
    }
}
```

Then you would have to write converter of some kind, logging framework dependent, here is example for logback:

```kotlin
class InteractiveConsoleConverter : CompositeConverter<ILoggingEvent>() {
    private val MOVE_CURSOR_TO_LINE_START = '\r'

    override fun transform(event: ILoggingEvent, input: String): String {
        if (input.contains(MOVE_CURSOR_TO_LINE_START)) {
            return MOVE_CURSOR_TO_LINE_START + input.replaceFirst(MOVE_CURSOR_TO_LINE_START, '\u0000').replace(System.lineSeparator(), "")
        }
        return event.message
    }
}
```

and wrap your wole pattern with something like:

```
%interactiveConsol([%d{YYYY-MM-dd HH:mm:ss.SSS}]%msg%n)
```

And yes, that actually works.

**Comment by o2themar** [comment]
@vehovsky Does the pull request that you created not require this work around. Or even with the new version I would still need to implement my own consumer?

**Comment by vehovsky** [comment]
@o2themar short answer, no. 

The pull requests ensures multiple progress bars can be displayed simultaneously, when supported by terminal in which you have started the JVM. 

When the terminal does not support moving cursor, now it at least ensures all "done" progressbars are displayed correctly. The "active" progressbars are re-writing each others at the line where cursos currently is (last line).

So you may want to create another issue for enhancing loggers integration.

**Comment by o2themar** [comment]
ok thanks @vehovsky I'll go with your suggestion for now about implementing a consumer to support the logging. I was just making sure that the new implantation didn't already cover this and you were suggestion a work around for now. Its not a big deal having to implement a consumer for the logging and I appreciate you sharing the code to help get me started. Thank you for your help in quickly getting a solution.

**Comment by ctongfei** [comment]
Fixed in `0.9.0` with @vehovsky 's fantastic addition.
